[WIP] re-enable solr-specific tests

icarito opened the pull request:
This is to replace https://github.com/publiclab/plots2/pull/1386 with a branch in plots2 repository in order to enable coveralls.

* [x] all tests pass -- `rake test:all`
* [x] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [x] pull request is descriptively named with #number reference back to original issue

icarito commented:
* [ ] figure out why Coveralls isn't running on here, and get it running so we can see if coverage drops - maybe by reopening this from `publiclab/plots2`?
* [ ] re-enable test.rake "solr-specific" tests in such a way as to not confuse Coveralls and maintain 86% coverage -- this may be due to those extra tests running in a subshell...
  * possibly with http://stackoverflow.com/questions/18894060/rails-and-minitest-add-additional-folder
  * possibly using alternative shell command runner: http://stackoverflow.com/questions/2232/calling-shell-commands-from-ruby#2400 
* [ ] fix tests on `test/solr/search_record_test.rb:50` to ensure unique results are actually returned
  * was this bc it was not actually indexing? So, maybe if we get Solr indexing properly, this will pass. 
* [ ] document why 23-hour reindex is needed, or fix the toggle to make this unnecessary
* [ ] re-enable solr on production

Then, the rest of this:

* [ ] get `branch1` building from this PR
* [ ] reindexing and then looking for if a search for "balloon" works on this page: http://branch1.laboratoriopublico.org/searches/new
* [ ] Once it does, we want to try turning off Solr and ensuring the site doesn't break (as the new toggle code should allow for).
* [ ] then, we can try merging in #1366, and seeing if http://branch1.laboratoriopublico.org/search/dynamic returns results for `balloon` with Solr on AND off
* [ ] then we have to do the remainder of #1366's listed to-dos

icarito commented:
Where is coveralls?...

icarito commented:
Great so we can check the coveralls issue now.

icarito commented:
Yay, figured out to fix the DISABLE_SOLR_CHECK (renamed) env var toggle to properly reindex in 47 minutes.

jywarren commented:
wow, cool!

On Wed, May 10, 2017 at 12:35 PM, Sebastian Silva <notifications@github.com>
wrote:

> Yay, figured out to fix the DISABLE_SOLR_CHECK (renamed) env var toggle to
> properly reindex in 47 minutes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/1409#issuecomment-300540260>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABfJ1JzP9qHkEz3sZjbfqXm7W6OHIIrks5r4edqgaJpZM4NWvJ2>
> .
>

icarito commented:
http://branch1.laboratoriopublico.org/searches/55
Searches feel slow to me, possibly from our pre-search solr check.

icarito commented:
> Once it does, we want to try turning off Solr and ensuring the site doesn't break (as the new toggle code should allow for).

It does break when I stop Solr. :-(

```
Started GET "/" for 190.232.146.3 at 2017-05-10 17:33:01 +0000
Processing by HomeController#home as HTML
Completed 500 Internal Server Error in 232.5ms

RSolr::Error::Http (RSolr::Error::Http - 503 Service Unavailable
Error:     {msg=SolrCore is loading,code=503}

URI: http://pad.publiclab.org:8983/solr/default/update?wt=ruby
Request Headers: {"Content-Type"=>"text/xml"}
Request Data: "<?xml version=\"1.0\" encoding=\"UTF-8\"?><delete><id>User 448589</id></delete>"

Backtrace: /usr/local/lib/ruby/gems/2.1.0/gems/rsolr-1.1.2/lib/rsolr/client.rb:288:in `adapt_response'
/usr/local/lib/ruby/gems/2.1.0/gems/rsolr-1.1.2/lib/rsolr/client.rb:189:in `execute'
/usr/local/lib/ruby/gems/2.1.0/gems/rsolr-1.1.2/lib/rsolr/client.rb:175:in `send_and_receive'
/usr/local/lib/ruby/gems/2.1.0/gems/sunspot_rails-2.2.6/lib/sunspot/rails/solr_instrumentation.rb:16:in `block in send_and_receive_with_as_instrumentation'
/usr/local/lib/ruby/gems/2.1.0/gems/activesupport-3.2.22.4/lib/active_support/notifications.rb:123:in `block in instrument'
/usr/local/lib/ruby/gems/2.1.0/gems/activesupport-3.2.22.4/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
/usr/local/lib/ruby/gems/2.1.0/gems/activesupport-3.2.22.4/lib/active_support/notifications.rb:123:in `instrument'
/usr/local/lib/ruby/gems/2.1.0/gems/sunspot_rails-2.2.6/lib/sunspot/rails/solr_instrumentation.rb:15:in `send_and_receive_with_as_instrumentation'
(eval):2:in `post'
/usr/local/lib/ruby/gems/2.1.0/gems/rsolr-1.1.2/lib/rsolr/client.rb:84:in `update'
/usr/local/lib/ruby/gems/2.1.0/gems/rsolr-1.1.2/lib/rsolr/client.rb:138:in `delete_by_id'):
  lib/authlogic_openid/authlogic_openid/acts_as_authentic.rb:74:in `save'
  app/controllers/application_controller.rb:71:in `current_user_session'
  app/controllers/application_controller.rb:76:in `current_user'
  app/controllers/home_controller.rb:13:in `home'
```

icarito commented:
>Once it does, we want to try turning off Solr and ensuring the site doesn't break (as the new toggle code should allow for).

Looks like it doesn't work, turning off Solr I got:
```
Started GET "/" for 190.232.146.3 at 2017-05-10 17:33:01 +0000
Processing by HomeController#home as HTML
Completed 500 Internal Server Error in 232.5ms

RSolr::Error::Http (RSolr::Error::Http - 503 Service Unavailable
Error:     {msg=SolrCore is loading,code=503}

URI: http://pad.publiclab.org:8983/solr/default/update?wt=ruby
Request Headers: {"Content-Type"=>"text/xml"}
Request Data: "<?xml version=\"1.0\" encoding=\"UTF-8\"?><delete><id>User 448589</id></delete>"

Backtrace: /usr/local/lib/ruby/gems/2.1.0/gems/rsolr-1.1.2/lib/rsolr/client.rb:288:in `adapt_response'
/usr/local/lib/ruby/gems/2.1.0/gems/rsolr-1.1.2/lib/rsolr/client.rb:189:in `execute'
/usr/local/lib/ruby/gems/2.1.0/gems/rsolr-1.1.2/lib/rsolr/client.rb:175:in `send_and_receive'
/usr/local/lib/ruby/gems/2.1.0/gems/sunspot_rails-2.2.6/lib/sunspot/rails/solr_instrumentation.rb:16:in `block in send_and_receive_with_as_instrumentation'
/usr/local/lib/ruby/gems/2.1.0/gems/activesupport-3.2.22.4/lib/active_support/notifications.rb:123:in `block in instrument'
/usr/local/lib/ruby/gems/2.1.0/gems/activesupport-3.2.22.4/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
/usr/local/lib/ruby/gems/2.1.0/gems/activesupport-3.2.22.4/lib/active_support/notifications.rb:123:in `instrument'
/usr/local/lib/ruby/gems/2.1.0/gems/sunspot_rails-2.2.6/lib/sunspot/rails/solr_instrumentation.rb:15:in `send_and_receive_with_as_instrumentation'
(eval):2:in `post'
/usr/local/lib/ruby/gems/2.1.0/gems/rsolr-1.1.2/lib/rsolr/client.rb:84:in `update'
/usr/local/lib/ruby/gems/2.1.0/gems/rsolr-1.1.2/lib/rsolr/client.rb:138:in `delete_by_id'):
  lib/authlogic_openid/authlogic_openid/acts_as_authentic.rb:74:in `save'
  app/controllers/application_controller.rb:71:in `current_user_session'
  app/controllers/application_controller.rb:76:in `current_user'
  app/controllers/home_controller.rb:13:in `home'
```

icarito commented:
This is the result of running solr tests in travis:
```
Command failed with status (1): [ruby -I"lib:test" -I"/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib" "/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/rake_test_loader.rb" "test/solr/search_record_test.rb" "test/solr/searches_controller_test.rb" ]
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/testtask.rb:109:in `block (3 levels) in define'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/file_utils.rb:57:in `call'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/file_utils.rb:57:in `sh'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/file_utils_ext.rb:37:in `sh'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/file_utils.rb:96:in `ruby'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/file_utils_ext.rb:37:in `ruby'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/testtask.rb:105:in `block (2 levels) in define'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/file_utils_ext.rb:58:in `verbose'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/testtask.rb:101:in `block in define'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/task.rb:240:in `call'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/task.rb:240:in `block in execute'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/task.rb:235:in `each'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/task.rb:235:in `execute'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/task.rb:179:in `block in invoke_with_call_chain'
/usr/local/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/task.rb:172:in `invoke_with_call_chain'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/task.rb:165:in `invoke'
/usr/local/lib/ruby/gems/2.1.0/gems/railties-3.2.22.4/lib/rails/test_unit/testing.rake:50:in `block in <top (required)>'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/task.rb:240:in `call'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/task.rb:240:in `block in execute'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/task.rb:235:in `each'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/task.rb:235:in `execute'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/task.rb:179:in `block in invoke_with_call_chain'
/usr/local/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/task.rb:172:in `invoke_with_call_chain'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/task.rb:165:in `invoke'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/application.rb:150:in `invoke_task'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/application.rb:106:in `block (2 levels) in top_level'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/application.rb:106:in `each'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/application.rb:106:in `block in top_level'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/application.rb:115:in `run_with_threads'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/application.rb:100:in `top_level'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/application.rb:78:in `block in run'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/application.rb:176:in `standard_exception_handling'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/application.rb:75:in `run'
/usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/bin/rake:33:in `<top (required)>'
/usr/local/bin/rake:23:in `load'
/usr/local/bin/rake:23:in `<main>'
Tasks: TOP => test:single
[Coveralls] Using SimpleCov's default settings.
Loaded suite /usr/local/lib/ruby/gems/2.1.0/gems/rake-10.5.0/lib/rake/rake_test_loader
Started
....E
===============================================================================
Error: test_should_get_search_test_action(SearchesControllerTest): JSON::ParserError: 757: unexpected token at 'null'
test/solr/searches_controller_test.rb:23:in `block in <class:SearchesControllerTest>'
===============================================================================
....
Finished in 2.07917866 seconds.
------
9 tests, 17 assertions, 0 failures, 1 errors, 0 pendings, 0 omissions, 0 notifications
0% passed
------
```

jywarren commented:
Huh, cool -- great work! But looks like the error at `test/solr/searches_controller_test.rb:23` is distinct, maybe new? Shall we add that to the checklist too?

jywarren commented:
But i got a bunch of email notifications about not working when Solr is off -- but the comments are gone now, was that resolved? Does it now work whether Solr is on or off?

icarito commented:
Yes sorry - i deleted those messages - initially I accidentally double posted. Then I removed one copy. Then I looked at a different tab not knowing it and saw the duplicate message, so I deleted it, but turns out I deleted both messages. Sigh.

Summarizing, yes, stopping Solr causes an error. :-/

icarito commented:
Here's the error we get when Solr is off:
```
Started GET "/searches/56" for 181.66.127.24 at 2017-05-12 15:17:55 +0000
Processing by SearchesController#show as HTML
  Parameters: {"id"=>"56"}
Completed 500 Internal Server Error in 67.8ms

RSolr::Error::ConnectionRefused (Connection refused - {:data=>"fq=type%3ANode&fq=updated_at_s%3A%7B*+TO+%222017%5C-05%5C-12%5C+15%5C%3A17%5C%3A55%5C+UTC%22%7D&q=balloon&fl=*+score&defType=edismax&start=0&rows=30&facet=true&f.updated_month_s.facet.mincount=1&facet.field=updated_month_s", :method=>:post, :params=>{:wt=>:ruby}, :query=>"wt=ruby", :headers=>{"Content-Type"=>"application/x-www-form-urlencoded; charset=UTF-8"}, :path=>"select", :uri=>#<URI::HTTP:0x007f6404245148 URL:http://pad.publiclab.org:8983/solr/default/select?wt=ruby>, :open_timeout=>nil, :read_timeout=>120, :retry_503=>nil, :retry_after_limit=>nil}):     app/models/search_record.rb:34:in `notes'                                                                                                                    app/controllers/searches_controller.rb:57:in `show'
                                                                                                                                                              
```

icarito commented:
Here's another coverage tool that perhaps even is used underneath coveralls?
https://github.com/colszowka/simplecov
It has pretty tables! ;-)

jywarren commented:
* Node.search is now being tested, could fail as of last couple commits.
* we are trying to reindex embedded Solr in Travis `rake test:solr`, but that might not work?
* it could need some sleep
* maybe ditch Coveralls for simplecov

jywarren commented:
Hmm, some green and red there, nice! 

BTW does Coveralls just use SimpleCov? https://travis-ci.org/publiclab/plots2/builds/231645129#L2782 says:

> [Coveralls] Using SimpleCov's default settings.

jywarren commented:
missed an `end` -- fixed it and added one more commit!

jywarren commented:
OK, reindex failed but because: `sh: 1: docker-compose: not found` on https://travis-ci.org/publiclab/plots2/builds/231657953#L2805

I think this is because we actually want to send it to embedded Solr, so we could just run `rake SOLR_DISABLE_CHECK=1 sunspot:reindex` without the docker compose stuff. Trying that.

jywarren commented:
Nice! Solr indexed: `[##################################] [72/72] [100.00%] [00:01] [00:00] [57.72/s]`

Trying to resolve another test failure now...

jywarren commented:
OK, in both cases it's not returning any results:

```

===============================================================================
Failure:
  <[]> expected to be != to
  <[]>.
test_should_get_search_test_action(SearchesControllerTest)
test/solr/searches_controller_test.rb:25:in `block in <class:SearchesControllerTest>'
===============================================================================
.F
===============================================================================
Failure: <false> is not true.
test_Node.search_for_two_different_key_words_returns_different_results(SearchRecordTest)
test/solr/search_record_test.rb:54:in `block in <class:SearchRecordTest>'
===============================================================================
```

However, `Node.search` does return results, so maybe we should just echo those out to be ...

OH! this is where we need htat extra parameter!

icarito commented:
Yay, making progress! Good work!


On 12/05/17 16:44, Jeffrey Warren wrote:
>
> OK, in both cases it's not returning any results:
>
> |===============================================================================
> Failure: <[]> expected to be != to <[]>.
> test_should_get_search_test_action(SearchesControllerTest)
> test/solr/searches_controller_test.rb:25:in `block in
> <class:SearchesControllerTest>'
> ===============================================================================
> .F
> ===============================================================================
> Failure: <false> is not true.
> test_Node.search_for_two_different_key_words_returns_different_results(SearchRecordTest)
> test/solr/search_record_test.rb:54:in `block in
> <class:SearchRecordTest>'
> ===============================================================================
> |
>
> However, |Node.search| does return results, so maybe we should just
> echo those out to be ...
>
> OH! this is where we need htat extra parameter!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/1409#issuecomment-301193158>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMMS5UY9vEhnidO7HgeWuNTRU7p5wvBks5r5NKhgaJpZM4NWvJ2>.
>

jywarren commented on test/solr/search_record_test.rb:
```diff
@@ -59,6 +59,8 @@ class SearchRecordTest < ActiveSupport::TestCase
       #with(:updated_month, month) if month.present?
       #paginate :page => 1, :per_page => 10
     end
+puts solr_search_1.results
```
Hmm, i have to wrap up for the day, but my attempt to output the results here didn't turn into anything. We could try `puts solr_search_1.results.inspect` instead...?

jywarren commented:
Some hints on coveralls vs. simplecov here: http://technology.indiegogo.com/tag/simplecov/ -- though we should maybe just break out a separate issue.

jywarren commented:
OK, after today's work, a checklist is:

* [ ] we're outputting all the 72 indexed nodes from the Solr auto-testing, to see why "Chicago" and "pectro" are not returning any records in search_record_test.rb
* [ ] @icarito is trying to get a rails console running with the test data so we can manually query it and see why it's not working in the tests.

We did confirm that in the staging server, these exact queries do return results, so the Solr test failures are confusing. All are related to `Node.search do ...` Solr queries.

icarito commented:
I ran the app with the "development" environment, and here's the logs from the Solr query for 'balloon' as constructed by the app versus constructed by the "Solr Admin".
```
solr_1  | 415450 INFO  (qtp1112280004-19) [   x:default] o.a.s.c.S.Request [default] webapp=/solr path=/select params={q=balloon&defType=edismax&facet.field=updated_month_s&fl=*+score&start=0&fq=type:Node&fq=updated_at_s:{*+TO+"2017\-06\-02\+05\:46\:27\+UTC"}&rows=30&wt=ruby&facet=true&f.updated_month_s.facet.mincount=1} hits=0 status=0 QTime=9 
solr_1  | 595278 INFO  (qtp1112280004-21) [   x:default] o.a.s.c.S.Request [default] webapp=/solr path=/select params={q=balloon&indent=true&wt=json&_=1496382567308} hits=848 status=0 QTime=2 
```
Notice the difference in hits. I bet the same happens in testing environment.

icarito commented:
Here's the log for a `irb` session with Solr in `test` RAILS_ENV:
```
irb(main):054:0> solr_search_1 = Node.search do
irb(main):055:1*   fulltext 'balloon'
irb(main):056:1>   #with(:updated_at).less_than(Time.zone.now)
irb(main):057:1*   # this is required to get results to return: 
irb(main):058:1*   adjust_solr_params do |params|
irb(main):059:2*     params[:qf] = nil
irb(main):060:2>   end
irb(main):061:1>    #facet(:updated_month)
irb(main):062:1*    #with(:updated_month, month) if month.present?
irb(main):063:1*    #paginate :page => 1, :per_page => 10
irb(main):064:1*  end
  SOLR Request (6.8ms)  [ path=select parameters={fq: ["type:Node"], q: "balloon", fl: "* score", qf: , defType: "edismax", start: 0, rows: 30} ]
=> <Sunspot::Search:{:fq=>["type:Node"], :q=>"balloon", :fl=>"* score", :qf=>nil, :defType=>"edismax", :start=>0, :rows=>30}>
irb(main):065:0> solr_search_1.results
=> []
```
Same thing with "Chicago" keyword.
Here's the corresponding Solr log line:
```
solr_1  | 711436 INFO  (qtp1112280004-18) [   x:default] o.a.s.c.S.Request [default] webapp=/solr path=/select params={q=balloon&defType=edismax&fl=*+score&start=0&fq=type:Node&rows=30&wt=ruby} hits=0 status=0 QTime=1 
```

icarito commented:
I think I see what is going on, look at the query URL:
`http://localhost:8982/solr/default/select?q=balloon&fq=*&wt=json&indent=true&fl=*+score&start=0&fq=type:Node&rows=30&wt=ruby`

If I start removing parameters I get results with removing `&fq=type:Node` . Switching `&fq=type:Node` for `&fq=type:DrupalNode` does make results. I wonder if test.sqlite is migrated?

icarito commented:
I see test.sqlite db has proper table names. I wonder where DrupalNode is coming from...?

jywarren commented:
When tests are run, the database is generated on the fly each time from the fixtures data, so `irb session with Solr in test RAILS_ENV` might not have a populated database -- have you double checked that there are records in it? And did a reindex happen based on the test db after it's been populated?

But i see below you do get results, so i guess it's set up right. Where else could DrupalNode be stored, is there a schema or something? It would have to be something tracked in Git if this is happening on TravisCI.

I just searched pretty thoroughly through the codebase and didn't find anything... so odd!

jywarren commented:
Running in test environment in a container, i ran `rails c` and got a console, then ran through:

```
> Node.search do
>   fulltext "Chicago"
>   adjust_solr_params do |params|
>     params[:qf] = nil
>   end
> end
  SOLR Request (15.8ms)  [ path=select parameters={fq: ["type:Node"], q: "Chicago", fl: "* score", qf: , defType: "edismax", start: 0, rows: 30} ]
RSolr::Error::Http: RSolr::Error::Http - 400 Bad Request
Error: {"responseHeader":{"status":400,"QTime":1,"params":{"q":"Chicago","defType":"edismax","fl":"* score","start":"0","fq":"type:Node","rows":"30","wt":"json"}},"error":{"msg":"undefined field type","code":400}}

URI: http://solr:8983/solr/default/select?wt=json
Request Headers: {"Content-Type"=>"application/x-www-form-urlencoded; charset=UTF-8"}
Request Data: "fq=type%3ANode&q=Chicago&fl=*+score&defType=edismax&start=0&rows=30"
```

jywarren commented:
So i was able to run:

```
rake sunspot:solr:start RAILS_ENV=test
RAILS_ENV=test rake SOLR_DISABLE_CHECK=1 sunspot:reindex
rails c
```

and then run `Node.search` just like in our tests in Travis, and get the same errors. Not sure why our fulltext searches aren't returning the records, since a plain `Node.search.results` does return all records, while a fulltext query does not.

jywarren commented:
WOOO we have to define the search fields, as per: https://github.com/sunspot/sunspot#full-text
```
# Posts with pizza in the title (boosted) or in the body (not boosted)
Post.search do
  fulltext 'pizza' do
    fields(:body, :title => 2.0)
  end
end
```

jywarren commented:
hmm, not sure why, but the searches aren't returning the results we'd seen in testing... i'm going to try to log back in and do it in a console in exactly the same circumstances.

jywarren commented:
OK, i found that the extra parameter tweak is not needed, and in fact breaks the query. Removed all instances.

jywarren commented:
Some minor fixtures tweaks. Getting very close now.

jywarren commented:
Well, what do you know? :-) 

Happy weekend!

jywarren commented:
Now we need to: 

* [ ] Once it does, we want to try turning off Solr and ensuring the site doesn't break (as the new toggle code should allow for).
* [ ] come up with a plan for solr failing on read -- that is, for `Node.search` when Solr is off
  * as distinct from on writing INTO solr. So, upon running a search. 
* [ ] re-enable solr on production
* [ ] then, we can try merging in #1366, and seeing if http://branch1.laboratoriopublico.org/search/dynamic returns results for `balloon` with Solr on AND off
* [ ] then we have to do the remainder of #1366's listed to-dos

icarito commented:
Great news! I guess the containerized Solr complicated everything. The Solr schema is still a question with embedded vs containerized Solr, possibly the reason the parameter needed tweaking in the first place. Thanks for advancing this!

jywarren commented:
Has this then built on branch1 so we can try it out?

On Jun 9, 2017 8:09 PM, "Sebastian Silva" <notifications@github.com> wrote:

> Great news! I guess the containerized Solr complicated everything. The
> Solr schema is still a question with embedded vs containerized Solr,
> possibly the reason the parameter needed tweaking in the first place.
> Thanks for advancing this!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/1409#issuecomment-307527831>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABfJ6b_W6--UXFLaG5bgClzaNJzhzPcks5sCd6pgaJpZM4NWvJ2>
> .
>

icarito commented:
It tried <http://jenkins.laboratoriopublico.org/job/Plots-Solr/96/> but
found issues (not sure why), I'll debug it and rebuild.



On 09/06/17 22:44, Jeffrey Warren wrote:
> Has this then built on branch1 so we can try it out?
>
> On Jun 9, 2017 8:09 PM, "Sebastian Silva" <notifications@github.com>
> wrote:
>
> > Great news! I guess the containerized Solr complicated everything. The
> > Solr schema is still a question with embedded vs containerized Solr,
> > possibly the reason the parameter needed tweaking in the first place.
> > Thanks for advancing this!
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/publiclab/plots2/pull/1409#issuecomment-307527831>,
> > or mute the thread
> >
> <https://github.com/notifications/unsubscribe-auth/AABfJ6b_W6--UXFLaG5bgClzaNJzhzPcks5sCd6pgaJpZM4NWvJ2>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/1409#issuecomment-307539515>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMMSx4nSUfHC0aaqqK2Xu0TSEsJLHt9ks5sChERgaJpZM4NWvJ2>.
>

jywarren commented:
Maybe related:

Removing network plots2_default network plots2_default has active endpoints
Build step 'Execute shell' marked build as failure


On Jun 10, 2017 8:37 AM, "Sebastian Silva" <notifications@github.com> wrote:

> It tried <http://jenkins.laboratoriopublico.org/job/Plots-Solr/96/> but
> found issues (not sure why), I'll debug it and rebuild.
>
>
>
> On 09/06/17 22:44, Jeffrey Warren wrote:
> > Has this then built on branch1 so we can try it out?
> >
> > On Jun 9, 2017 8:09 PM, "Sebastian Silva" <notifications@github.com>
> > wrote:
> >
> > > Great news! I guess the containerized Solr complicated everything. The
> > > Solr schema is still a question with embedded vs containerized Solr,
> > > possibly the reason the parameter needed tweaking in the first place.
> > > Thanks for advancing this!
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/publiclab/plots2/pull/1409#issuecomment-307527831
> >,
> > > or mute the thread
> > >
> > <https://github.com/notifications/unsubscribe-auth/AABfJ6b_W6--
> UXFLaG5bgClzaNJzhzPcks5sCd6pgaJpZM4NWvJ2>
> > > .
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/publiclab/plots2/pull/1409#issuecomment-307539515>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AAMMSx4nSUfHC0aaqqK2Xu0TSEsJLHt9ks5sChERgaJpZM4NWvJ2>.
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/1409#issuecomment-307562632>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABfJzCFaUomJZu_H01UFD1uraaOXp1Dks5sCo3ZgaJpZM4NWvJ2>
> .
>

icarito commented:
I manually removed orphan cobtainers with docker-compose down
--remove-orphans and added this to the jenkins config. Its reindexing now i
think.


El sábado, 10 de junio de 2017, Jeffrey Warren <notifications@github.com>
escribió:

> Maybe related:
>
> Removing network plots2_default network plots2_default has active endpoints
> Build step 'Execute shell' marked build as failure
>
>
> On Jun 10, 2017 8:37 AM, "Sebastian Silva" <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
> > It tried <http://jenkins.laboratoriopublico.org/job/Plots-Solr/96/> but
> > found issues (not sure why), I'll debug it and rebuild.
> >
> >
> >
> > On 09/06/17 22:44, Jeffrey Warren wrote:
> > > Has this then built on branch1 so we can try it out?
> > >
> > > On Jun 9, 2017 8:09 PM, "Sebastian Silva" <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>>
> > > wrote:
> > >
> > > > Great news! I guess the containerized Solr complicated everything.
> The
> > > > Solr schema is still a question with embedded vs containerized Solr,
> > > > possibly the reason the parameter needed tweaking in the first place.
> > > > Thanks for advancing this!
> > > >
> > > > —
> > > > You are receiving this because you were mentioned.
> > > > Reply to this email directly, view it on GitHub
> > > > <https://github.com/publiclab/plots2/pull/1409#issuecomment-
> 307527831
> > >,
> > > > or mute the thread
> > > >
> > > <https://github.com/notifications/unsubscribe-auth/AABfJ6b_W6--
> > UXFLaG5bgClzaNJzhzPcks5sCd6pgaJpZM4NWvJ2>
> > > > .
> > > >
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/publiclab/plots2/pull/1409#issuecomment-307539515
> >,
> > > or mute the thread
> > > <https://github.com/notifications/unsubscribe-auth/
> > AAMMSx4nSUfHC0aaqqK2Xu0TSEsJLHt9ks5sChERgaJpZM4NWvJ2>.
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/publiclab/plots2/pull/1409#issuecomment-307562632>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AABfJzCFaUomJZu_
> H01UFD1uraaOXp1Dks5sCo3ZgaJpZM4NWvJ2>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/1409#issuecomment-307562920>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMMS2IjI0qsEo1NG-EIaaownp0jnmygks5sCo9OgaJpZM4NWvJ2>
> .
>


-- 
Sebastian Silva
http://somosazucar.org/

icarito commented:
So I checked and branch1.laboratoriopublico.org works as expected, I will be checking failsafe mechanism in the weekend.

jywarren commented:
Hm, though I can't seem to load our test search: http://branch1.laboratoriopublico.org/searches/test

That's supposed to run a Solr search according to:

https://github.com/publiclab/plots2/blob/master/config/routes.rb#L19 and 
https://github.com/publiclab/plots2/blob/master/app/controllers/searches_controller.rb#L11

Can you check the logs to see why that's returning a 404?

jywarren commented:
Whoa all right: http://branch1.laboratoriopublico.org/searches/test/balloon

That's a solr search!

jywarren commented:
And to search for arbitrary terms: 

http://branch1.laboratoriopublico.org/searches/test/test?q=balloon
http://branch1.laboratoriopublico.org/searches/test/test?q=kite

These don't return anything, though. Maybe branch1 needs to reindex the initial time? It looks like it's running off the test data.

jywarren commented:
Yes, well, see; http://branch1.laboratoriopublico.org/searches/test/test?q=Spectrometer

icarito commented:
Thanks for keeping looking at this. The data in branch1 is from a mysqldump from 20170330.
I will look at what happens when I stop the Solr container.

icarito commented:
Here's an excerpt from the logs, getting 500 error when Solr is off, in Home and Dashboard:
```
web_1  |                                                                                                                                                     │
web_1  | Started GET "/" for 200.106.34.51 at 2017-06-10 19:17:44 +0000                                                                                      │
web_1  | Processing by HomeController#home as HTML                                                                                                           │
web_1  | Completed 500 Internal Server Error in 274.8ms                                                                                                      │
web_1  |                                                                                                                                                     │
web_1  | RSolr::Error::ConnectionRefused (Connection refused - {:data=>"<?xml version=\"1.0\" encoding=\"UTF-8\"?><delete><id>User 448589</id></delete>", :he│
aders=>{"Content-Type"=>"text/xml"}, :method=>:post, :params=>{:wt=>:ruby}, :query=>"wt=ruby", :path=>"update", :uri=>#<URI::HTTP:0x007f48009f8c70 URL:http:/│
/pad.publiclab.org:8983/solr/default/update?wt=ruby>, :open_timeout=>nil, :read_timeout=>120, :retry_503=>nil, :retry_after_limit=>nil}):                    │
web_1  |   lib/authlogic_openid/authlogic_openid/acts_as_authentic.rb:74:in `save'                                                                           │
web_1  |   app/controllers/application_controller.rb:71:in `current_user_session'                                                                            │
web_1  |   app/controllers/application_controller.rb:76:in `current_user'                                                                                    │
web_1  |   app/controllers/home_controller.rb:13:in `home'                                                                                                   │
web_1  |                                                                                                                                                     │
web_1  |                                                                                                                                                     │
web_1  | Started GET "/dashboard" for 200.106.34.51 at 2017-06-10 19:17:46 +0000                                                                             │
web_1  | Processing by HomeController#dashboard as HTML                                                                                                      │
web_1  | Completed 500 Internal Server Error in 292.4ms                                                                                                      │
web_1  |                                                                                                                                                     │
web_1  | RSolr::Error::ConnectionRefused (Connection refused - {:data=>"<?xml version=\"1.0\" encoding=\"UTF-8\"?><delete><id>User 448589</id></delete>", :he│
aders=>{"Content-Type"=>"text/xml"}, :method=>:post, :params=>{:wt=>:ruby}, :query=>"wt=ruby", :path=>"update", :uri=>#<URI::HTTP:0x007f48006e3198 URL:http:/│
/pad.publiclab.org:8983/solr/default/update?wt=ruby>, :open_timeout=>nil, :read_timeout=>120, :retry_503=>nil, :retry_after_limit=>nil}):                    │
web_1  |   lib/authlogic_openid/authlogic_openid/acts_as_authentic.rb:74:in `save'                                                                           │
web_1  |   app/controllers/application_controller.rb:71:in `current_user_session'                                                                            │
web_1  |   app/controllers/application_controller.rb:76:in `current_user'                                                                                    │
web_1  |   app/controllers/home_controller.rb:64:in `dashboard'                                                                                              │
web_1  |                                                                                                                                                     │
web_1  |              
```

jywarren commented:
OK, yes this is the need for a "toggle for Solr reading, not just writing" -- but just checking, has the index been generated on the entire db you mentioned, or is it possible that it's been overwritten with an index based on the test data?

icarito commented:
It is definitively the reindexed data from the mysqldump we're using in
branch1. Jenkins is set to reindex upon successful built:


    Statistics

Last Modified:
    a day ago
Num Docs:
    829529



On 12/06/17 11:55, Jeffrey Warren wrote:
>
> OK, yes this is the need for a "toggle for Solr reading, not just
> writing" -- but just checking, has the index been generated on the
> entire db you mentioned, or is it possible that it's been overwritten
> with an index based on the test data?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/1409#issuecomment-307850107>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMMS39GeeJpIdvyOUCzmEeeAPt6gqSQks5sDW2EgaJpZM4NWvJ2>.
>

jywarren commented:
Hmm, it seems to be down now: http://branch1.laboratoriopublico.org/searches/test/test?q=balloon

So next step is to add a check to that controller to ensure the solr toggle is on before executing the search. We can test this toggle on the `test` method here, and the expand it to other instances of `Node.search`: https://github.com/publiclab/plots2/blob/master/app/controllers/searches_controller.rb#L11

jywarren commented:
OK, I attempted this in 1c356f7

jywarren commented:
Hmm, apparently you can't `include` from a controller? OK, i'm trying without the `include` since we are [auto-loading paths](https://github.com/publiclab/plots2/blob/reenable-solr-tests/config/application.rb#L23)

icarito commented:
> Hmm, it seems to be down now: http://branch1.laboratoriopublico.org/searches/test/test?q=balloon

Interesting. Actually, it's up. I use a ssh tunnel to be able to access the Solr Admin on pad.publiclab.org (it's firewalled otherwise, except from tycho). In case you were going to ask, I create the tunnel with GSTM, a gui frontend, but `ps aux | grep ssh` reveals the command it makes for this: `ssh pad.publiclab.org -p 22 -l icarito -nN -L8983:localhost:8983`

jywarren commented:
Ah sorry i meant the Rails app, not Solr. It's back up now, maybe it was down for rebuilding/indexing. It's showing a Rails error so I'll keep debugging -- probably same as the error shown on Travis about `include` in a controller...

icarito commented:
Ah, okay, yes, the Jenkins job is set to stop the service first, then
build, then start it, then reindex.


On 14/06/17 10:17, Jeffrey Warren wrote:
>
> Ah sorry i meant the Rails app, not Solr. It's back up now, maybe it
> was down for rebuilding/indexing. It's showing a Rails error so I'll
> keep debugging -- probably same as the error shown on Travis about
> |include| in a controller...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/1409#issuecomment-308464643>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMMS8tJzUmj4o6Tyj0h3P5crZYo8kEEks5sD_migaJpZM4NWvJ2>.
>

jywarren commented:
Rebased and attempted a fix

jywarren commented:
aha -- hmm, this time, we dropped coverage again! how/why!?

jywarren commented:
Awesome, this worked perfectly! 

`Solr search service offline`

We can write a wrapper for Node.search which checks this first, such as Node.search_if_available!

jywarren commented:
@icarito, if I add this, would you be ready to reenable Solr on production?

icarito commented:
GREAT! pad.publiclab.org is ready with a Solr container.
I believe pushing to staging should work cleanly :-)
:trophy: :palm_tree: :fireworks: :1st_place_medal: :grinning:

icarito commented:
Looking forward to put this in production. :-)

jywarren commented:
Fingers X, i've implemented on-read toggles for all instances of `Node.search`.

jywarren commented:
We could test this out on Sunday if you like -- assuming nothing else comes up. I want to be very thorough about testing with Solr on and off, in staging.

jywarren commented:
And in production actually!

jywarren commented:
Wheeee!

jywarren commented:
Nice, and both http://branch1.laboratoriopublico.org/ and http://branch1.laboratoriopublico.org/searches/test/test?q=About load normally -- with Solr off. I think this is ready for a merge!

jywarren commented on docker-compose.yml:
```diff
@@ -13,16 +13,16 @@ services:
-      - docker-entrypoint.sh
-      - solr-precreate
-      - default
+#  solr:
```
Also, just checking -- do we need to re-enable this?

jywarren replied:
also any relation to https://github.com/publiclab/plots2/pull/1428 ?

icarito replied:
Not in production or staging as we're planning to deploy the solr container in pad.publiclab.org.

icarito replied:
But I'll add a separate docker-compose-production.yml for this, so we might as well reenable this here.

jywarren commented:
Do you want to make the required changes, as I'm not 100% sure I know --
and maybe add a comment accordingly inline? Then i'm ready to merge :-)

On Wed, Jun 14, 2017 at 8:30 PM, Sebastian Silva <notifications@github.com>
wrote:

> *@icarito* commented on this pull request.
> ------------------------------
>
> In docker-compose.yml
> <https://github.com/publiclab/plots2/pull/1409#discussion_r122096516>:
>
> > @@ -13,16 +13,16 @@ services:
>        - ../dump:/docker-entrypoint-initdb.d
>        # for importing a dump
>        # if you want to re-import simply put away ../mysql/*
> -  solr:
> -    image: solr:5.3
> -    ports:
> -      - "127.0.0.1:8984:8983"
> -    volumes:
> -      - solr:/opt/solr/server/solr/mycores
> -    entrypoint:
> -      - docker-entrypoint.sh
> -      - solr-precreate
> -      - default
> +#  solr:
>
> But I'll add a separate docker-compose-production.yml for this, so we
> might as well reenable this here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/1409#discussion_r122096516>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABfJ2gxHbUqYxJCrd0lg-rbA3n06D89ks5sEHtBgaJpZM4NWvJ2>
> .
>

icarito commented:
It should work as-is, if I understood correctly. I'm going to be afk for the next few hours.
Regards,
Sebastian

jywarren commented:
Ah, so i can merge this?

icarito commented:
Yes, we can merge the commented version, I believe. Sunspot.yml doesn't point to the `solr` container but to localhost, and it's off by default as Solr is only optional. I'm adding an explanatory comment, though.

Just also add a note for the sake of thoroughness, the Solr schema creation process (which we assumed occurred during `rake db:setup`) showed differences when done with the embedded engine vs the container (to the point where queries required different parameters). It would be good to take these into consideration when putting in production, and even better, to understand why.

jywarren commented:
Ok!!!

icarito commented:
Oh wow that was fast I was committing the additional comment on docker-compose.yml -
